raphtory/src/db/graph/graph.rs

```diff
@@ -726,6 +729,337 @@ mod db_tests {
         assert_eq!(gg.edges().len(), 2);
     }
 
+    #[test]
+    fn import_node_as() {
+        let g = Graph::new();
+        let g_a = g.add_node(0, "A", NO_PROPS, None).unwrap();
+        let g_b = g
+            .add_node(1, "B", vec![("temp".to_string(), Prop::Bool(true))], None)
+            .unwrap();
+        let _ = g_b.add_constant_properties(vec![("con".to_string(), Prop::I64(11))]);
+
+        let gg = Graph::new();
+        let res = gg.import_node_as(&g_a, "X", false).unwrap();
+        assert_eq!(res.name(), "X");
+        assert_eq!(res.history(), vec![0]);
+
+        let _ = gg.add_node(1, "Y", NO_PROPS, None).unwrap();
+        let res = gg.import_node_as(&g_b, "Y", false);
+        match res {
+            Err(NodeExistsError(id)) => {
+                assert_eq!(
+                    id.to_string(),
+                    "Y",
+                    "Unexpected node ID for NodeExistsError"
+                );
+            }
+            Err(e) => panic!("Unexpected error: {:?}", e),
+            Ok(_) => panic!("Expected error but got Ok"),
+        }
+
+        let mut nodes = gg.nodes().name().collect_vec();
+        nodes.sort();
+        assert_eq!(nodes, vec!["X", "Y"]); // Nodes up until first failure are imported
+        let y = gg.node("Y").unwrap();
+
+        assert_eq!(y.name(), "Y");
+        assert_eq!(y.history(), vec![1]);
+        assert_eq!(y.properties().get("temp"), None);
+        assert_eq!(y.properties().constant().get("con"), None);
+    }
+
+    #[test]
+    fn import_node_as_force() {
+        let g = Graph::new();
+        let g_a = g.add_node(0, "A", NO_PROPS, None).unwrap();
+        let g_b = g
+            .add_node(1, "B", vec![("temp".to_string(), Prop::Bool(true))], None)
+            .unwrap();
+        let _ = g_b.add_constant_properties(vec![("con".to_string(), Prop::I64(11))]);
+
+        let gg = Graph::new();
+        gg.add_node(1, "Y", NO_PROPS, None).unwrap();
+
+        let res = gg.import_node_as(&g_a, "X", false).unwrap();
+        assert_eq!(res.name(), "X");
+        assert_eq!(res.history(), vec![0]);
+
+        let res = gg.import_node_as(&g_b, "Y", true).unwrap();
+        assert_eq!(res.name(), "Y");
+        assert_eq!(res.history(), vec![1]);
+        assert_eq!(res.properties().get("temp").unwrap(), Prop::Bool(true));
+        assert_eq!(
+            res.properties().constant().get("con").unwrap(),
+            Prop::I64(11)
+        );
+    }
+
+    #[test]
+    fn import_nodes_as() {
+        let g = Graph::new();
+        let g_a = g.add_node(0, "A", NO_PROPS, None).unwrap();
+        let g_b = g
+            .add_node(1, "B", vec![("temp".to_string(), Prop::Bool(true))], None)
+            .unwrap();
+        let _ = g_b.add_constant_properties(vec![("con".to_string(), Prop::I64(11))]);
+
+        let gg = Graph::new();
+        gg.add_node(1, "Q", NO_PROPS, None).unwrap();
+        let res = gg.import_nodes_as(vec![&g_a, &g_b], vec!["P", "Q"], false);
```

Comment:
I think we would want this to either fail from the beginning (and import none) or to return a list of failed imports such that you can rename them